javascript/computer_science/project_binary_search_trees.md

```diff
@@ -0,0 +1,46 @@
+### Introduction
+
+You have learned about [binary search trees](http://en.wikipedia.org/wiki/Binary_search_tree) -- where you take a group of data items and turn them into a tree full of nodes where each left node is "lower" than each right node.  The tree starts with the "root node" and any node with no children is called a "leaf node". You have also learned about tree traversal algorithms like breadth-first and depth-first.
+
+Now, let's take a look at balanced binary search trees (BST). Read [this article](https://www.geeksforgeeks.org/sorted-array-to-balanced-bst/) and watch [this video](https://youtu.be/VCTP81Ij-EM) to understand the basic algorithm used to build a balanced BST. Although the last resource does not use Javascript, you should understand it enough to develop your own pseudocode.
+
+### Assignment
+You'll build a balanced BST in this assignment. Do not use duplicate values because they make it more complicated and result in trees that are much harder to balance. Therefore, be sure to always remove duplicate values or check for an existing value before inserting.
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Build a `Node` object.  It should have an attribute for the data it stores as well as its left and right children.
+
+1. Build a `Tree` object which accepts an array when initialized. The `Tree` object should have a `root` attribute which uses the return value of `build_tree` which you'll write next.
+
+1. Write a `build_tree` function which takes an array of data (e.g. [1, 7, 4, 23, 8, 9, 4, 3, 5, 7, 9, 67, 6345, 324]) and turns it into a balanced binary tree full of `Node` objects appropriately placed (don't forget to sort and remove duplicates!). The `build_tree` function should return the level-0 root node.
+
+1. Write an `insert` and `delete` functions which accepts a value to insert/delete (you'll have to deal with several cases for delete such as when a node has children or not). If you need additional resources, check out these two articles on [inserting](https://www.geeksforgeeks.org/binary-search-tree-set-1-search-and-insertion/?ref=lbp) and [deleting](https://www.geeksforgeeks.org/binary-search-tree-set-2-delete/?ref=lbp), or [this video](https://youtu.be/wcIRPqTR3Kc) with several visual examples.
+
+1. Write a `find` function which accepts a value and returns the node with the given value.
+
+1. Write a `level_order` function which accepts another function as a parameter. `level_order` should traverse the tree in breadth-first level order and yield each node to the provided function. This function can be implemented using either iteration or recursion (try implementing both!). The method should return an array of values if no function is given. **Tip:** You will want to use an array acting as a queue to keep track of all the child nodes that you have yet to traverse and to add new ones to the list (as you saw in the [video](https://www.youtube.com/watch?v=86g8jAQug04)).
```

Comment:
Yield each node to the provided function sounds weird for JS. Wording suggestion:
```suggestion
1. Write a `levelOrder` function which accepts another function as a parameter. `levelOrder` should traverse the tree in breadth-first level order and provide each node as the argument to the provided function. This function can be implemented using either iteration or recursion (try implementing both!). The method should return an array of values if no function is given. **Tip:** You will want to use an array acting as a queue to keep track of all the child nodes that you have yet to traverse and to add new ones to the list (as you saw in the [video](https://www.youtube.com/watch?v=86g8jAQug04)).
```